### PR TITLE
fix: Use Circle ci branch filter and remove master-ci-branch.sh

### LINKED ISSUE #6024: Use CircleCI's built-in branch filtering feature throughout
CircleCI supports [branch filters with regular expressions](https://circleci.com/docs/2.0/configuration-reference/#jobfilters). We're currently using the script [`match-ci-branch.sh`](https://github.com/freedomofpress/securedrop/blob/develop/devops/scripts/match-ci-branch.sh) for our branch filter logic.

We should evaluate if there are still reasons to use this script, and if not, replace it with CircleCI's built-in functionality, for performance and maintainability reasons.

### PR DESCRIPTION
**Pull request by prateekj117**
## Status

Ready for review

## Description of Changes

Fixes #6024 

Changes proposed in this pull request:
Circle CI supports [branch filters with regular expressions](https://circleci.com/docs/2.0/configuration-reference/#jobfilters). Right now, we are using both the [master-ci-branch.sh](https://github.com/freedomofpress/securedrop/blob/develop/devops/scripts/match-ci-branch.sh) as well as circle CI filters. This PR removes the usage of the script with circle CI filters throughout. 

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation
- [X] These changes do not require documentation

### CODE REVIEW

File: .circleci/config.yml
Comment (prateekj117) [inline .circleci/config.yml]: @eloquence I am not sure if these changes are valid as these jobs only run on develop branch weekly, and `deb-tests` and `translation-tests` wouldn't run by these filters. But I made these changes as earlier while using the `match-ci-branch.sh` script, these filters used to apply anyway, and I was not sure if we wanted a change in that behavior or not.

File: .circleci/config.yml
Comment (cfm) [inline .circleci/config.yml]: I appreciate your including the `ignore` list here, as both (a) a record of past behavior and (b) an indicator of what the desired behavior would be without the `only` list taking precedence.  My only concern is that this may not be obvious to folks reading this file for the first time without knowledge of this history (#5336) , who may be tempted to remove the `ignore` list in a way that's confusing down the line.  Would you mind adding a comment to this effect?

File: .circleci/config.yml
Comment (cfm) [reply .circleci/config.yml]: I agree that the `branches` specifications can be removed for jobs under the `weekly` workflow too.  Now that [CircleCI has deprecated `branches` specifications at the job level](https://circleci.com/docs/2.0/configuration-reference/#branches-deprecated), these would only be relevant if these jobs were added to the `securedrop_ci` or some other per-branch workflow.  The `weekly` configuration will be much simpler without them!

File: .circleci/config.yml
Comment (cfm) [reply .circleci/config.yml]: Nit:  Different whitespace on lines 356 versus 358.  If you like, `ignore` and `only` could each be just [a string rather than a single-item list of strings](https://circleci.com/docs/2.0/configuration-reference/#branches), though both conventions are used in this file.  :-)

File: .circleci/config.yml
Comment (cfm) [inline .circleci/config.yml]: Since the `nightly` workflow runs only on `develop`, I don't think it's necessary to ignore any branches for the `staging-test-with-rebase` job as part of that workflow.

**Review by cfm — CHANGES_REQUESTED** [review]
Thanks, @prateekj117!  I've reviewed the diff and agree with how you've translated the `match-ci-branch.sh` calls into CircleCI's `branches` configuration.

- [x] Tests ran on this branch as expected with this CircleCI configuration.

I've left a few comments and nits for the sake of simplicity and readability, but then this looks good for a final maintainer's review!

**Comment by eloquence** [comment]
@prateekj117 Just touching base, do you have time to continue your work on this?

**Comment by prateekj117** [comment]
@eloquence Just caught up in some office work. Will do these changes in a day or two.

**Comment by prateekj117** [comment]
@eloquence @cfm Ready for another review.

**Review by cfm — DISMISSED** [review]
Thanks for revising, @prateekj117!  Looks correct and tidy to me.  The CI failure in `app-page-layout-tests` is due to #6104 and unrelated to these changes.

Approved for final maintainer's review.

**Review by conorsch — APPROVED** [review]
Rebased on top of latest develop. Fine to merge once CI passes!

**Pull request merged**
